server/src/test/java/org/opensearch/index/compositeindex/datacube/startree/builder/AbstractStarTreeBuilderTests.java

```diff
@@ -1091,6 +1204,192 @@ private static Iterator<StarTreeDocument> getExpectedStarTreeDocumentIterator()
+
+        int noOfStarTreeDocuments2 = 7;
+        StarTreeDocument[] starTreeDocuments2 = new StarTreeDocument[noOfStarTreeDocuments2];
+        starTreeDocuments2[0] = new StarTreeDocument(new Long[] { 1L, 11L, 21L }, new Double[] { 400.0 });
+        starTreeDocuments2[1] = new StarTreeDocument(new Long[] { 1L, 12L, 22L }, new Double[] { 200.0 });
+        starTreeDocuments2[2] = new StarTreeDocument(new Long[] { 2L, 13L, 23L }, new Double[] { 300.0 });
+        starTreeDocuments2[3] = new StarTreeDocument(new Long[] { 2L, 13L, 21L }, new Double[] { 100.0 });
+        starTreeDocuments2[4] = new StarTreeDocument(new Long[] { 3L, 11L, 21L }, new Double[] { 600.0 });
+        starTreeDocuments2[5] = new StarTreeDocument(new Long[] { 3L, 12L, 23L }, new Double[] { 200.0 });
+        starTreeDocuments2[6] = new StarTreeDocument(new Long[] { 3L, 12L, 21L }, new Double[] { 400.0 });
+
+        StarTreeDocument[] segmentStarTreeDocuments2 = new StarTreeDocument[noOfStarTreeDocuments2];
+        for (int i = 0; i < noOfStarTreeDocuments2; i++) {
+            long metric1 = NumericUtils.doubleToSortableLong((Double) starTreeDocuments2[i].metrics[0]);
+            segmentStarTreeDocuments2[i] = new StarTreeDocument(starTreeDocuments2[i].dimensions, new Long[] { metric1 });
+        }
+
+        SequentialDocValuesIterator[] dimsIterators2 = getDimensionIterators(segmentStarTreeDocuments2);
+        List<SequentialDocValuesIterator> metricsIterators2 = getMetricIterators(segmentStarTreeDocuments2);
+        builder = getStarTreeBuilder(metaOut, dataOut, compositeField, writeState, mapperService);
+        Iterator<StarTreeDocument> segmentStarTreeDocumentIterator2 = builder.sortAndAggregateSegmentDocuments(
+            dimsIterators2,
+            metricsIterators2
+        );
+        builder.build(segmentStarTreeDocumentIterator2, new AtomicInteger(), mock(DocValuesConsumer.class));
+        InMemoryTreeNode rootNode2 = builder.getRootNode();
+
+        metaOut.close();
+        dataOut.close();
+
+        // length of the data is 1303 bytes. 16 bytes for header. Each Node is 33 bytes. We have 39 nodes ~ 33 * 39 = 1287. Total = 1303.
+        StarTreeMetadata starTreeMetadata = new StarTreeMetadata(
+            "test",
+            STAR_TREE,
+            mock(IndexInput.class),
+            List.of("field1", "field3", "field5", "field8"),
+            List.of(
+                new MetricEntry("field2", MetricStat.SUM),
+                new MetricEntry("field4", MetricStat.SUM),
+                new MetricEntry("field6", MetricStat.COUNT),
+                new MetricEntry("field9", MetricStat.MIN),
+                new MetricEntry("field10", MetricStat.MAX)
+            ),
+            2,
+            1,
+            Set.of("field8"),
+            getBuildMode(),
+            0,
+            346
+        );
+
+        // length of the data is 1303 bytes. 16 bytes for header. Each Node is 33 bytes. We have 39 nodes ~ 33 * 39 = 1287. Total = 1303.
```

Comment:
nit : Refactor, pull to separate methods

Reply:
same comment as below resolving